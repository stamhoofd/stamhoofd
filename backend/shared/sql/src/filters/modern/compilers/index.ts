export * from './less';
export * from './greater';
export * from './equals';
export * from './in';
