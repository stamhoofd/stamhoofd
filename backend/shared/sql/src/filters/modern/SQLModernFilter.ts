import { SimpleError } from '@simonbackx/simple-errors';
import { compileFilter, FilterCompiler, FilterDefinitions, filterDefinitionsToCompiler, RequiredFilterCompiler, StamhoofdFilter } from '@stamhoofd/structures';
import { SQLExpression, SQLExpressionOptions, SQLQuery } from '../../SQLExpression';
import { SQLJoin } from '../../SQLJoin';
import { SQLJsonValue } from '../../SQLJsonExpressions';
import { SQLSelect } from '../../SQLSelect';
import { SQLWhere, SQLWhereAnd, SQLWhereExists, SQLWhereJoin, SQLWhereNot, SQLWhereOr } from '../../SQLWhere';
import { $equalsSQLFilterCompiler, $greaterThanSQLFilterCompiler, $inSQLFilterCompiler, $lessThanSQLFilterCompiler } from './compilers';
import { $containsSQLFilterCompiler } from './compilers/contains';

export type SQLSyncFilterRunner = (column: SQLCurrentColumn) => SQLWhere;
export type SQLFilterRunner = (column: SQLCurrentColumn) => Promise<SQLWhere> | SQLWhere;
export type SQLFilterCompiler = FilterCompiler<SQLFilterRunner>;
export type SQLRequiredFilterCompiler = RequiredFilterCompiler<SQLFilterRunner>;
export type SQLFilterDefinitions = FilterDefinitions<SQLFilterRunner>;

export enum SQLValueType {
    /** At the root of a select */
    Table = 'Table',

    /** Column with type string */
    String = 'String',

    /** MySQL Datetime */
    Datetime = 'Datetime',

    /** Column with type number */
    Number = 'Number',

    /** Column with type boolean, meaning 1 or 0 */
    Boolean = 'Boolean',

    /** True or false in JSON */
    JSONBoolean = 'JSONBoolean',
    JSONString = 'JSONString',

    JSONNumber = 'JSONNumber',

    /** [...] */
    JSONArray = 'JSONArray',

    /** {...} */
    JSONObject = 'JSONObject',
}

export type SQLCurrentColumn = {
    expression: SQLExpression;

    /**
     * MySQL nullable. Please fill this in correctly! If a value can be null, can not exist (=mysql null), or can be JSONNull, set this to true.
     *
     * Mainly > and < operators will make sure the behaviour is consistent with MySQL sorting (normally comparing with null will always return false in MySQL)
     */
    nullable?: boolean;

    /**
     * JSON nullable
     */

    /**
     * Type of this column, use to normalize values received from filters
     */
    type: SQLValueType;
    checkPermission?: () => Promise<void>;
};

export function createColumnFilter(column: SQLCurrentColumn, childDefinitions?: SQLFilterDefinitions): SQLFilterCompiler {
    return (filter: StamhoofdFilter) => {
        const compiler = childDefinitions ? filterDefinitionsToCompiler(childDefinitions) : filterDefinitionsToCompiler(baseSQLFilterCompilers);
        const runner = $andSQLFilterCompiler(filter, compiler);

        return async (_: SQLCurrentColumn) => {
            if (column.checkPermission) {
                await column.checkPermission();
            }
            return await runner({
                nullable: false,
                ...column,
            });
        };
    };
}

export function createWildcardColumnFilter(getColumn: (key: string) => SQLCurrentColumn, childDefinitions?: (key: string) => SQLFilterDefinitions): SQLFilterCompiler {
    const wildcardCompiler = (filter: StamhoofdFilter, _, key: string) => {
        const compiler = childDefinitions ? filterDefinitionsToCompiler(childDefinitions(key)) : filterDefinitionsToCompiler(baseSQLFilterCompilers);
        const runner = $andSQLFilterCompiler(filter, compiler);

        return async (_: SQLCurrentColumn) => {
            const column = getColumn(key);
            if (column.checkPermission) {
                await column.checkPermission();
            }
            return await runner({
                nullable: false,
                ...column,
            });
        };
    };

    return (filter: StamhoofdFilter) => {
        return $andSQLFilterCompiler(filter, wildcardCompiler);
    };
}

/**
 * Filter with a subquery that should return at least one result.
 */
export function createExistsFilter(baseSelect: InstanceType<typeof SQLSelect> & SQLExpression, definitions: SQLFilterDefinitions): SQLFilterCompiler {
    return (filter: StamhoofdFilter, _: SQLFilterCompiler) => {
        if (filter !== null && typeof filter === 'object' && '$elemMatch' in filter) {
            filter = filter['$elemMatch'] as StamhoofdFilter;
        }

        const runner = compileToSQLRunner(filter, definitions);

        return async (_: SQLCurrentColumn) => {
            const w = await runner({
                expression: SQLRootExpression,
                type: SQLValueType.Table,
                nullable: false,
            });
            const q = baseSelect.clone().andWhere(w);
            return new SQLWhereExists(q);
        };
    };
}

/**
 * WARNING: only use this on one-to-one relations. Using it on one-to-many relations will result in duplicate results.
 *
 * By default doesRelationAlwaysExist is set to true, this means we expect the relation to always exist. This helps optimize the query (dropping the join if the where clause in the join is always true)
 */
export function createJoinedRelationFilter(join: SQLJoin, definitions: SQLFilterDefinitions, options: { doesRelationAlwaysExist: boolean } = { doesRelationAlwaysExist: true }): SQLFilterCompiler {
    return (filter: StamhoofdFilter, _: SQLFilterCompiler) => {
        if (filter !== null && typeof filter === 'object' && '$elemMatch' in filter) {
            filter = filter['$elemMatch'] as StamhoofdFilter;
        }

        return async (_: SQLCurrentColumn) => {
            const w = await compileToModernSQLFilter(filter, definitions);
            return new SQLWhereJoin(join, w, {
                doesRelationAlwaysExist: options.doesRelationAlwaysExist,
            });
        };
    };
}

export function $andSQLFilterCompiler(filter: StamhoofdFilter, filters: SQLFilterCompiler): SQLFilterRunner {
    const runners = compileSQLFilter(filter, filters);

    return async (column: SQLCurrentColumn) => {
        const wheres = (await Promise.all(
            runners.map(runner => (runner(column))),
        ));

        return new SQLWhereAnd(wheres);
    };
}

export function $orSQLFilterCompiler(filter: StamhoofdFilter, filters: SQLFilterCompiler): SQLFilterRunner {
    const runners = compileSQLFilter(filter, filters);

    return async (column: SQLCurrentColumn) => {
        const wheres = (await Promise.all(
            runners.map(runner => (runner(column))),
        ));

        return new SQLWhereOr(wheres);
    };
}

export function $notSQLFilterCompiler(filter: StamhoofdFilter, filters: SQLFilterCompiler): SQLFilterRunner {
    const andRunner = $andSQLFilterCompiler(filter, filters);

    return async (column: SQLCurrentColumn) => {
        return new SQLWhereNot(await andRunner(column));
    };
}

function invertFilterCompiler(compiler: SQLRequiredFilterCompiler): SQLRequiredFilterCompiler {
    return (filter: StamhoofdFilter, parentCompiler: SQLFilterCompiler) => {
        const runner = compiler(filter, parentCompiler);
        return async (column) => {
            return new SQLWhereNot(await runner(column));
        };
    };
}

export const baseSQLFilterCompilers: SQLFilterDefinitions = {
    $and: $andSQLFilterCompiler,
    $or: $orSQLFilterCompiler,
    $not: $notSQLFilterCompiler,
    $eq: $equalsSQLFilterCompiler,
    $neq: invertFilterCompiler($equalsSQLFilterCompiler),

    $lt: $lessThanSQLFilterCompiler,
    $gt: $greaterThanSQLFilterCompiler,
    $lte: invertFilterCompiler($greaterThanSQLFilterCompiler),
    $gte: invertFilterCompiler($lessThanSQLFilterCompiler),

    $in: $inSQLFilterCompiler,

    $contains: $containsSQLFilterCompiler,
};

const compileSQLFilter = compileFilter<SQLFilterRunner>;

export const SQLRootExpression: SQLExpression = {
    getSQL(options?: SQLExpressionOptions): SQLQuery {
        throw new SimpleError({
            code: 'invalid_filter',
            message: 'Root level filters are not allowed to use $eq or $neq',
        });
    },
};

export function compileToSQLRunner(filter: StamhoofdFilter, definitions: SQLFilterDefinitions): SQLFilterRunner {
    if (filter === null) {
        return () => {
            return new SQLWhereAnd([]); // No filter, return empty where
        };
    }
    const compiler = filterDefinitionsToCompiler(definitions); // this compiler searches in the definition for the right compiler for the given key
    const runner = $andSQLFilterCompiler(filter, compiler);
    return runner;
};

export async function compileToModernSQLFilter(filter: StamhoofdFilter, filters: SQLFilterDefinitions): Promise<SQLWhere> {
    const runner = compileToSQLRunner(filter, filters);
    return await runner({
        expression: SQLRootExpression,
        type: SQLValueType.Table,
        nullable: false,
    });
};

/**
 * Casts json strings, numbers and booleans to native MySQL types. This includes json null to mysql null.
 */
export function normalizeColumn(column: SQLCurrentColumn): SQLCurrentColumn {
    if (column.type === SQLValueType.JSONString) {
        return {
            expression: new SQLJsonValue(column.expression, 'CHAR'),
            type: SQLValueType.String,
        };
    }

    if (column.type === SQLValueType.JSONBoolean) {
        return {
            expression: new SQLJsonValue(column.expression, 'UNSIGNED'),
            type: SQLValueType.Boolean,
        };
    }

    if (column.type === SQLValueType.JSONNumber) {
        return {
            expression: new SQLJsonValue(column.expression, 'UNSIGNED'),
            type: SQLValueType.Number,
        };
    }
    return column;
}
