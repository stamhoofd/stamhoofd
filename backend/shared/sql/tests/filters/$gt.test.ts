import { baseModernSQLFilterCompilers, createColumnFilter, SQLValueType } from '../../src/filters/modern/SQLModernFilter';
import { SQL } from '../../src/SQL';
import { TableDefinition, test, testMatch } from '../utils';

describe('$gt', () => {
    /**
     * Tests that should be repeated for all filter types
     */
    describe('Common checks', () => {
        // todo
    });

    it('can filter on $gt', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $gt: 18,
                },
            },
            filters,
            query: {
                query: '`default`.`age` > ?',
                params: [18],
            },
        });
    });

    it('can invert $gt inside', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $not: {
                        $gt: 18,
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` <= ?',
                params: [18],
            },
        });
    });

    it('can invert $gt outside', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                $not: {
                    age: {
                        $gt: 18,
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` <= ?',
                params: [18],
            },
        });
    });

    it('can compare string values with $gt', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            name: createColumnFilter({ expression: SQL.column('name'), type: SQLValueType.String, nullable: false }),
        };
        await test({
            filter: {
                name: {
                    $gt: 'Alice',
                },
            },
            filters,
            query: {
                query: '`default`.`name` > ?',
                params: ['alice'],
            },
        });
    });

    it('can compare JSON string values with $gt', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            'settings.name': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.name'), type: SQLValueType.JSONString, nullable: false },
            ),
        };
        await test({
            filter: {
                'settings.name': {
                    $gt: 'Alice',
                },
            },
            filters,
            query: {
                query: 'JSON_VALUE(`default`.`settings`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?',
                params: ['alice'],
            },
        });
    });

    describe('MySQL Behaviour', () => {
        const tableDefinition: TableDefinition = {
            settings: {
                type: 'json',
                nullable: true,
            },
            name: {
                type: 'varchar',
                nullable: true,
            },
            age: {
                type: 'int',
                nullable: true,
            },
        };
        const filters = {
            ...baseModernSQLFilterCompilers,
            'name': createColumnFilter(
                { expression: SQL.column('name'), type: SQLValueType.String, nullable: true },
            ),
            'age': createColumnFilter(
                { expression: SQL.column('age'), type: SQLValueType.Number, nullable: true },
            ),
            'settings.name': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.name'), type: SQLValueType.JSONString, nullable: true },
            ),
            'settings': createColumnFilter({ expression: SQL.column('settings'), type: SQLValueType.JSONObject, nullable: true }),
            'settings.enabled': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.enabled'), type: SQLValueType.JSONBoolean, nullable: true },
            ),
            'settings.age': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.age'), type: SQLValueType.JSONNumber, nullable: true },
            ),
        };

        it('Comparing string values is case insensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        name: 'alice',
                    },
                    {
                        name: 'Alice',
                    },
                    {
                        name: 'ALICE',
                    },
                ],
                doMatch: [
                    {
                        name: {
                            $gt: 'a',
                        },
                    },
                    {
                        name: {
                            $gt: 'alic',
                        },
                    },
                    {
                        name: {
                            $gt: 'Alic',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        name: {
                            $gt: 'Alice',
                        },
                    },
                    {
                        name: {
                            $gt: 'aLice',
                        },
                    },
                    {
                        name: {
                            $gt: 'ALICE',
                        },
                    },
                    {
                        name: {
                            $gt: 'alice',
                        },
                    },
                    {
                        name: {
                            $gt: 'b',
                        },
                    },
                ],
            });
        });

        it('Comparing JSON string values is case insensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: 'alice',
                        },
                    },
                    {
                        settings: {
                            name: 'Alice',
                        },
                    },
                    {
                        settings: {
                            name: 'ALICE',
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.name': {
                            $gt: 'a',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'alic',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'Alic',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.name': {
                            $gt: 'Alice',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'aLice',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'ALICE',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'alice',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'b',
                        },
                    },
                ],
            });
        });

        it('Comparing numbers values is numberic', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        age: 10,
                    },
                ],
                doMatch: [
                    {
                        age: {
                            $gt: 9,
                        },
                    },
                    {
                        age: {
                            $gt: 0,
                        },
                    },
                    {
                        age: {
                            $gt: -9,
                        },
                    },
                    {
                        age: {
                            $gt: 1,
                        },
                    },
                ],
                doNotMatch: [
                    {
                        age: {
                            $gt: 10,
                        },
                    },
                    {
                        age: {
                            $gt: 12,
                        },
                    },
                    {
                        age: {
                            $gt: 20,
                        },
                    },
                ],
            });
        });

        it('Comparing JSON numbers values is numberic', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            age: 10,
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.age': {
                            $gt: 9,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: 0,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: -9,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: 1,
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.age': {
                            $gt: 10,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: 12,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: 20,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for JSON numbers', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            age: null,
                        },
                    },
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                ],
                doNotMatch: [
                    {
                        'settings.age': {
                            $gt: -1000,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: 0,
                        },
                    },
                    {
                        'settings.age': {
                            $gt: null,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for JSON strings', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: null,
                        },
                    },
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                ],
                doNotMatch: [
                    {
                        'settings.name': {
                            $gt: '',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: 'aaaa',
                        },
                    },
                    {
                        'settings.name': {
                            $gt: null,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for strings', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        name: null,
                    },
                ],
                doNotMatch: [
                    {
                        name: {
                            $gt: '',
                        },
                    },
                    {
                        name: {
                            $gt: 'aaaa',
                        },
                    },
                    {
                        name: {
                            $gt: null,
                        },
                    },
                ],
            });
        });
    });
});
