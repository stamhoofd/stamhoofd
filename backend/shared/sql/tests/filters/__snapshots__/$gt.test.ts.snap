// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$gt":-9}} 1`] = `
{
  "params": [
    -9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$gt":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$gt":1}} 1`] = `
{
  "params": [
    1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$gt":9}} 1`] = `
{
  "params": [
    9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$gt":10}} 1`] = `
{
  "params": [
    10,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$gt":12}} 1`] = `
{
  "params": [
    12,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$gt":20}} 1`] = `
{
  "params": [
    20,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for filter: {"settings.name":{"$gt":"Alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for filter: {"settings.name":{"$gt":"a"}} 1`] = `
{
  "params": [
    "a",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for filter: {"settings.name":{"$gt":"alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for not match filter: {"settings.name":{"$gt":"ALICE"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for not match filter: {"settings.name":{"$gt":"Alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for not match filter: {"settings.name":{"$gt":"aLice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for not match filter: {"settings.name":{"$gt":"alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing JSON string values is case insensitive: SQL Query for not match filter: {"settings.name":{"$gt":"b"}} 1`] = `
{
  "params": [
    "b",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$gt":-9}} 1`] = `
{
  "params": [
    -9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$gt":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$gt":1}} 1`] = `
{
  "params": [
    1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$gt":9}} 1`] = `
{
  "params": [
    9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$gt":10}} 1`] = `
{
  "params": [
    10,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$gt":12}} 1`] = `
{
  "params": [
    12,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$gt":20}} 1`] = `
{
  "params": [
    20,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$gt":"Alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$gt":"a"}} 1`] = `
{
  "params": [
    "a",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$gt":"alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$gt":"ALICE"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$gt":"Alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$gt":"aLice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$gt":"alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$gt":"b"}} 1`] = `
{
  "params": [
    "b",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for not match filter: {"settings.age":{"$gt":-1000}} 1`] = `
{
  "params": [
    -1000,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for not match filter: {"settings.age":{"$gt":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for not match filter: {"settings.age":{"$gt":null}} 1`] = `
{
  "params": [],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.age" RETURNING UNSIGNED ERROR ON ERROR) IS NOT NULL",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for not match filter: {"settings.name":{"$gt":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for not match filter: {"settings.name":{"$gt":"aaaa"}} 1`] = `
{
  "params": [
    "aaaa",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for not match filter: {"settings.name":{"$gt":null}} 1`] = `
{
  "params": [],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) IS NOT NULL",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for strings: SQL Query for not match filter: {"name":{"$gt":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for strings: SQL Query for not match filter: {"name":{"$gt":"aaaa"}} 1`] = `
{
  "params": [
    "aaaa",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` > ?",
}
`;

exports[`$gt MySQL Behaviour NULL values are always the smallest for strings: SQL Query for not match filter: {"name":{"$gt":null}} 1`] = `
{
  "params": [],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NOT NULL",
}
`;
